### ISSUE TITLE
`gensquashfs@1.3.1` fails to pack directory with `glob`

### ISSUE THREAD

**Issue by birunts** [AgentD/squashfs-tools-ng#129]
Having example directory structure:

```
└ misc
  ├── build
  │   ├── subbuild
  │   │   └── subtool.py
  │   └── test.sh
  ├── test_a.txt
  └── test_b.txt
```
and a `packfile`:
```
glob /a/b/c 0775 0 0 -type d -- misc
glob /a/b/c 0775 0 0 -type f -name "*" -- misc
```
Trying to pack `misc` directory content to `/a/b/c` fails with:
```
$ gensquashfs --pack-file packfile --pack-dir=$(pwd) output.squashfs
packing a/b/c/build/subbuild/subtool.py
a/b/c/build/subbuild/subtool.py: No such file or directory
```

The problem is with [populate_dir()](https://github.com/AgentD/squashfs-tools-ng/blob/fixes-1.3.0/bin/gensquashfs/fstree_from_dir.c#L284) function which takes output name as an input file.

Looks like [tree_node_t.data.file.input_file](https://github.com/AgentD/squashfs-tools-ng/blob/fixes-1.3.0/include/fstree.h#L114) is not set. This would most probably need set [extra](https://github.com/AgentD/squashfs-tools-ng/blob/fixes-1.3.0/bin/gensquashfs/fstree_from_dir.c#L288) for `S_ISREG == true`.

**Comment by AgentD** [comment]
Hi @birunts ,

yes the field `input_file` in `tree_node_t` stores the full source path, if applicable. When we scan a directory, or the pack file does not specify a source path (i.e. same path as on-disk), we can skip that and instead reconstruct the path in `populate_dir()` from the tree node.

The problem in your case is that we scanning a directory hierarchy under a tree node that isn't root (here `a/b/c`), but the scanning code fails to test for that. We then end up with broken paths when packing the files, because the path only matches up to that root node.

On current master, the problem has apparently been fixed while refactoring the scanning code.

I managed to fix your test case with the 1.3.x branch by patching the directory scanning code, to store the full path if the root node is not the root of the tree.

Here is the diff: [file_prefix.diff.gz](https://github.com/user-attachments/files/15952652/file_prefix.diff.gz)

I hope this works for you? I will try to reconcile this with the Windows port, do some more testing and get another patch level release out soon.

Thanks for reporting!

**Comment by birunts** [comment]
Hi @AgentD,

thank you for sending the patch. This is _more or less_ what I actually did to fix it locally.

However your patch seems not to work as expected. The difference is that you are concatenating the path to the prefix, which I actually don't understand why is that. 

Here is my quick patch [squashfs-tools-ng_fix.patch.tar.gz](https://github.com/user-attachments/files/16232009/squashfs-tools-ng_fix.patch.tar.gz) which is working for me (no changes to things using `gensquashfs`, same as before with [9f2d631](https://github.com/AgentD/squashfs-tools-ng/commit/9f2d63162e82d423faeda9c7449d6edd819e4591) I used).

Would you please have a look at this and verify correctness of your patch? I do not know `gensquashfs` specifics so most probable  I misunderstood sth. 

Thanks!

**Comment by birunts** [comment]
@AgentD have you had a chance to look at this? Any updates? 

Thanks in advance :slightly_smiling_face:

**Comment by AgentD** [comment]
Hi @birunts ,

I'm sorry the issue stalled for so long. I've been a little busier than expected over the last 2+ months of "U-boot duty" (BSP/bring-up on 3 boards with project deadlines that Wham! once made a famous song about) and also managed to build up a nasty backlog of mtd-utils patches.

Anyway, I did take a few cursory looks at your patch and did some side-by-side comparisons. The approaches *do* look very similar, with the difference that I tried to pull the path concatenation logic out into a separate function, to not touch `populate_dir` too much.

You said that my patch does not work as expected? I did test it with the test case you provided. Is it packing/not-packing something as expected?

I guess a main difference might be in the way the prefix is passed into `fstree_from_subdir`. There is additional logic in `gensquashfs` to `chdir` into a base directory (either where the manifest file lives, to make the paths relative to it, or the pack directory). Maybe I tested my version in a different way and this is where the discrepancy is from?
